.github/workflows/ci.yml

```diff
@@ -63,8 +63,58 @@ jobs:
       - name: run integration tests
         run: ./scripts/integration.sh
 
+  gpu-folders:
+    runs-on: [ self-hosted, gpu ]
+    steps:
+      - name: Change dir owner to working user
+        run: sudo chown -R $USER:$USER $GITHUB_WORKSPACE
+      - name: Checkout
+        uses: actions/checkout@v2
+      - id: changed-files
+        uses: jitterbit/get-changed-files@v1
+        continue-on-error: true
+      - id: set-matrix
+        run: bash scripts/get-gpu-executors.sh
+        env:
+          CHANGED_FILES: ${{ steps.changed-files.outputs.all }}
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.matrix }}
+
+  gpu-tests:
+    needs: [ gpu-folders ]
+    runs-on: [ self-hosted, gpu ]
+    strategy:
+      fail-fast: false
+      matrix:
+        folder: ${{fromJson(needs.gpu-folders.outputs.matrix)}}
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: debug
+        run: echo ${{ matrix.folder }}
+      - name: build docker image
+        run: docker build -t $(echo "$(basename "${{ matrix.folder }}"):gpu" | tr '[:upper:]' '[:lower:]') -f ${{ matrix.folder }}/Dockerfile.gpu ${{ matrix.folder }}
+      - name: start container
+        run: |
+          docker run \
+            -v "$(pwd):/test_workspace" \
+            --entrypoint "" \
+            --workdir /test_workspace \
+            --name $(basename ${{ matrix.folder }}) \
+            --gpus all \
+            -d $(echo "$(basename "${{ matrix.folder }}"):gpu" | tr '[:upper:]' '[:lower:]') tail -F /dev/null
+          export DOCKER_EXEC="docker exec $(basename "${{ matrix.folder }}")"
+      - name: run tests in container
+        run: |
+          $DOCKER_EXEC ls
+          $DOCKER_EXEC pip install -r tests/requirements.txt
+          $DOCKER_EXEC pytest -s -v -m gpu tests/
+      - name: stop container
+        if: ${{ always() }}
+        run: docker stop $(docker ps -a -q --filter="name=$(basename "${{ matrix.folder }}")")
```

Comment:
I think we don't need the filter - as this is the only job running on the runner, I think it is safe to kill all containers.